README.md
# ap_scanner


## Command Line Access Point (AP) scanning tool 


### <u>Overview:</u>
This python utility will scan for network Access Points (APs) using underlying OS utilities and list related AP information.

- Supports Linux and Windows. 
- Output options include: formatted (default), csv and json


### <u>Install:</u>
```
pip install ap_scanner
```

or for source code:

```
git clone https://github.com/JavaWiz1/ap_scanner.git
```

Note: poetry is used for dependency management and packaging

### <u>Usage:</u>
```
usage: ap_scanner [-h] [-i <iface>] [-r] [-j] [-c] [-v] [--nmcli] [--iwlist] [--iw] [--netsh]

Scan for wi-fi access points (Networks)

options:
  -h, --help            show this help message and exit
  -i <iface>, --interface <iface>
                        (Linux only) Interface to use, default=wlan0
  -r, --rescan          (Windows only) force network rescan for APs
  -j, --json            Output json result
  -c, --csv             Output csv result
  -v, --verbose         Debug/verbose output to console
  --nmcli               Force Linux Network Manager discovery
  --iwlist              Force Linux iwlist discovery
  --iw                  Force Linux iw discover
  --netsh               Force Windows netsh discovery

This utility will scan for network APs (Access Points) using underlying OS utilities
and list related information.

- Supports Linux and Windows.
- Output options include: formatted (default), csv and json
```


### <u>Examples:</u>
Windows - trigger re-scan to get most current list of access points
```
> ap_scanner -r
===========================================================
==                   ap_scanner v0.1.1                   ==
===========================================================
==        Scan for wi-fi access points (Networks)        ==
===========================================================

Validate command line options
- 1 Wifi adapter(s) detected: Wi-Fi
- "Wi-Fi" will be used to scan via 802.11ac [99%]
- Windows Scanner netsh selected

Rescan requested
- Autoconnect enabled for MyLAN1
- Disconnect to trigger re-scan of network

Scan for access points (WindowsWiFiScanner)
- Executing: netsh wlan show network mode=bssid

Process results of scan
- 5 APs discovered

SSID                      Auth            Encryption Mac Address       Signal Radio    Band    Channel
------------------------- --------------- ---------- ----------------- ------ -------- ------- -------
MyLAN1                    WPA2-Personal   CCMP       0c:9a:91:2c:bb:28   99%  802.11n  Unknown       3
                                                     0c:9a:91:2c:bb:2c   99%  802.11ac Unknown      36
NETGEAR99                 WPA2-Personal   CCMP       b0:41:a0:85:0f:d6   99%  802.11n  Unknown       3
**hidden**                WPA2-Enterprise CCMP       dc:eb:62:0a:b0:8b   33%  802.11n  Unknown       1
                                                     42:f1:9a:6d:5c:7b   21%  802.11n  Unknown       1
                                                     22:47:5c:fa:01:39   18%  802.11ac Unknown       6
Nirvana                   Open            None       f4:cf:b2:ac:bd:b1   58%  802.11n  Unknown       1
ThePond                   WPA2-Personal   CCMP       7c:80:26:01:70:de   51%  802.11n  Unknown       1
```

Linux - Use wlan2 connection for scan and use Network Manager (nmcli) for discovery
```
> ap_scanner -i wlan2 --nmcli
===========================================================
==                   ap_scanner v0.1.1                   ==
===========================================================
==        Scan for wi-fi access points (Networks)        ==
===========================================================

Validate command line options
- 1 Wifi adapter(s) detected: wlan2
- "wlan2" will be used to scan

Scan for access points (NetworkManagerWiFiScanner)
- Executing: /usr/bin/nmcli -t -f ssid,bssid,chan,freq,signal,security,rsn-flags device wifi list

Process results of scan
- 5 APs discovered

SSID                      Auth            Encryption Mac Address       Signal Radio    Band    Channel
------------------------- --------------- ---------- ----------------- ------ -------- ------- -------
MyLAN1                    WPA2-Personal   CCMP       0C-9A-91-2C-BB-28  100%  Unknown  2.4 MHz       3
NETGEAR99                 WPA2-Personal   CCMP       B0-41-A0-85-0F-D6  100%  Unknown  2.4 MHz       3
**hidden**                WPA2-Personal   CCMP       88-9E-62-1D-3E-1A  100%  Unknown  2.4 MHz      11
Nirvana                   Open            None       F4-CF-B2-AC-BD-B1   44%  Unknown  2.4 MHz       1
ThePond                   WPA2-Personal   CCMP       7C-80-26-01-70-DE   44%  Unknown  2.4 MHz       1
```

### <u>Notes:</u>
- On windows, you can rescan (-r) and search for current networks
- You may save the outut of the underlying command into a file (-s)
- You can force which method searches for networks:
  - Linux:   nmcli, iw, iwlist
  - Windows: netsh
- Tested on
  - Windows 10/11
  - Ubuntu Ubuntu 22.04.3 LTS
  - RaspPi OS - Debian GNU/Linux 11 (bullseye) / Raspbian GNU/Linux 10 (buster)


### <u>ToDo:</u>
- Identify Band (i.e 2.4 MHz / 5 Mhz) on Windows
- Radio identification for Linux (802.11xx)
- Add Apple MAC capability
- Create unit tests

CHANGELOG.md
ChangeLog for ap_scanner
====================================================================================
v0.1.2 | 10/02/23 | short description
- Fix pyproject.toml to include changelog and supporting modules
- Update README.md to reflect current output
- Remove scapy dependency (not used)

====================================================================================

v0.1.1 | 10/02/23 | Restructure / Cleanup
- no functionality changes
- repackage code modules
- documentation strings and logging
- *** THIS VERSION IS BROKEN - do not use ***

====================================================================================

v0.1.0 | 10/01/23 | Initial release  
- Support for Windows/Linux
- Tested on 
    - Windows 10/11
    - Ubuntu Ubuntu 22.04.3 LTS
    - RaspPi OS - Debian GNU/Linux 11 (bullseye) / Raspbian GNU/Linux 10 (buster)



utils.py
import json
import platform
from typing import List

from loguru import logger as LOGGER

from scanner import CONSTANTS, AccessPoint, ScannerBase


def running_on_linux() -> bool:
    """Return True if running on Linux OS, else False"""
    return platform.system() == "Linux"

def running_on_windows() -> bool:
    """Return True if running on Windows OS, else False"""
    return platform.system() == "Windows"


def os_check(scanner: ScannerBase) -> bool:
    """Return True if running on supported OS, else False"""
    if running_on_windows() and scanner.scanner_supported_os() == CONSTANTS.WINDOWS:
        return True
    elif running_on_linux() and scanner.scanner_supported_os() == CONSTANTS.LINUX:
        return True

    return False

def get_input(prompt: str, valid_responses: list = [], default: str = None) -> str:
    """
    Prompt for valid input
    Parameters:
        prompt          - req - text to display
        valid_responses - opt - stringco or list of valid responses (default None)
        default         - opt - default vault returned (default None)
    """
    valid_input = False
    while not valid_input:
        response = input(prompt)
        if not valid_responses:
            LOGGER.debug('no valid responses to check')
            valid_input = True
        else:
            if response in valid_responses:
                valid_input = True

    return response


def to_dict(obj, classkey=None):
    """Recursively translate object into dictionary format"""
    if isinstance(obj, dict):
        data = {}
        for (k, v) in obj.items():
            data[k] = to_dict(v, classkey)
        return data
    elif hasattr(obj, "_ast"):
        return to_dict(obj._ast())
    elif hasattr(obj, "__iter__") and not isinstance(obj, str):
        return [to_dict(v, classkey) for v in obj]
    elif hasattr(obj, "__dict__"):
        data = dict([(key, to_dict(value, classkey)) 
            for key, value in obj.__dict__.items() 
            if not callable(value) and not key.startswith('_')])
        if classkey is not None and hasattr(obj, "__class__"):
            data[classkey] = obj.__class__.__name__
        return data
    else:
        return obj

#===========================================================================================================================
#== Output routines
def display_access_points(ap_list: List[AccessPoint]):
    """Display formatted AccessPoint output"""
    LOGGER.info('')
    LOGGER.info('SSID                      Auth            Encryption Mac Address       Signal Radio    Band    Channel')
    LOGGER.info('------------------------- --------------- ---------- ----------------- ------ -------- ------- -------')
    for sidx in range(len(ap_list)):
        ap = ap_list[sidx]
        bssid = ap.bssid[0]
        LOGGER.info(f'{ap.ssid.name:25} {ap.ssid.auth:15} {ap.ssid.encryption:10} {bssid.mac:17} {bssid.signal:4}%  {bssid.radio_type:8} {bssid.band:7} {bssid.channel:7}')
        for bidx in range(1, len(ap.bssid)):
            bssid = ap.bssid[bidx]
            LOGGER.info(f'{" "*52} {bssid.mac:17} {bssid.signal:4}%  {bssid.radio_type:8} {bssid.band:7} {bssid.channel:7}')

def display_json(ap_list: List[AccessPoint]):
    """Display AccessPoint output in json format"""
    LOGGER.info('- json output')
    print(json.dumps(to_dict(ap_list),indent=2))

def display_csv(ap_list: List[AccessPoint]):
    """Display AccessPoint output in csv format"""
    LOGGER.info('- csv output')
    print('ssid,auth,encryption,mac,signal,type,band,channel')
    for ap in ap_list:
        ssid_info = f'{ap.ssid.name},{ap.ssid.auth},{ap.ssid.encryption}'
        for bssid in ap.bssid:
            bssid_info = f'{bssid.mac},{bssid.signal},{bssid.radio_type},{bssid.band},{bssid.channel}'
            print(f'{ssid_info},{bssid_info}')



ap_scanner.py
import argparse
import sys
from dataclasses import dataclass
from importlib.metadata import version
from typing import List

from loguru import logger as LOGGER

from utils import *
from scanner import WindowsWiFiScanner, IwWiFiScanner, IwlistWiFiScanner, NetworkManagerWiFiScanner, ScannerBase
from scanner import CONSTANTS


# == Adapter Object ==========================================================================================================   
@dataclass 
class AdapterInfo:
    name: str
    desc: str = ''
    mac: str = ''
    connected: bool = False
    SSID: str = 'Not Associated'
    BSSID: str = ''
    radio_type: str = ''
    Authentication: str = ''
    cipher: str = ''
    channel: int = -1
    receive_rate: float = -1.0
    transmit_rate: float = -1.0
    signal: int = -1

    def __post_init__(self):
        if running_on_windows():
            if not self._get_windows_adapter():
                raise NameError(f'Unable to identify adapter "{self.name}"')
        elif running_on_linux():
            if not self._get_linux_adapter():
                raise NameError(f'Unable to identify adapter "{self.name}"')
        else:
            raise NameError(f'No adapter exists named "{self.name}"')
        
    def _get_linux_adapter(self) -> bool:
        # iw wlan0 info = channel, mhz, mac
        iwconfig_output, _ = ScannerBase._execute_process('iwconfig', show_feedback=False)
        adapter_found = False
        for line in iwconfig_output:
            line = line.strip()
            if 'ESSID' in line:
                if adapter_found:
                    # done, bail
                    break
                adapter_found = True
                self.SSID = line.split('ESSID:')[1].strip()
                continue
            if "Frequency:" in line:
                token = line.split('Frequency:')[1].strip()
                if token.startswith('2'):
                    self.radio_type = "2.4 GHz"
                elif token.startswith('5'):
                    self.radio_type = '5 GHz'
            if "Access Point:" in line:
                self.BSSID = line.split('Access Point:')[1].strip()
            if "Bit Rate=" in line:
                self.receive_rate = line.split('Bit Rate=')[1].split()[0]
                self.transmit_rate = self.receive_rate
            if "Link Quality=" in line:
                txt_sig = line.split('=')[1].split()[0]
                signals = txt_sig.split('/')
                self.signal = int(int(signals[0]) / int(signals[1]) * 100)
            # self.Authentication
            # self.channel
            # self.cipher
            # self.connected
            # self.desc
            # self.mac
            # self.radio_type

        return adapter_found
    
    def _get_windows_adapter(self) -> bool:
        netsh_output, _ = ScannerBase._execute_process(f'netsh wlan show interfaces', show_feedback=False)
        adapter_found = False
        for line in netsh_output:
            line = line.strip()
            if line.startswith('Name'):
                if adapter_found:
                    # Fully processed target name, quit
                    break
                else:
                    adapter_found = True
                    continue
            value = '' if len(line.split(':',1)) == 1 else line.split(':',1)[1].strip()
            if line.startswith('Description'):
                self.desc = value
                continue
            if line.startswith('Physical address'):
                self.mac = value
                continue
            if line.startswith('State'):
                self.connected = value == 'connected'
                continue
            if line.startswith('SSID'):
                self.SSID = value
                continue
            if line.startswith('BSSID'):
                self.BSSID = value
                continue
            if line.startswith('Radio type'):
                self.radio_type = value
                continue
            if line.startswith('Authentication'):
                self.Authentication = value
                continue
            if line.startswith('Cipher'):
                self.cipher = value
                continue
            if line.startswith('Channel'):
                self.channel = int(value)
                continue
            if line.startswith('Receive rate'):
                self.receive_rate = float(value)
                continue
            if line.startswith('Transmit rate'):
                self.transmit_rate = float(value)
                continue
            if line.startswith('Signal'):
                self.signal = int(value.replace('%',''))
                continue

        return adapter_found
    

# ============================================================================================================================
# == Helper routines =========================================================================================================
def identify_scanner(args: argparse.Namespace) -> ScannerBase:
    """
    Return scanner object based on:
    1. command line parameter (if supplied)
    2. OS and installed utilties in order (nmcli, iwlist, iw)
    """
    scanner: ScannerBase = None
    if args.nmcli:  
        scanner = NetworkManagerWiFiScanner(interface=args.interface)
        LOGGER.info('- Scanner nmcli requested (Linux)')
    elif args.iwlist:
        scanner = IwlistWiFiScanner(interface=args.interface)
        LOGGER.info('- Scanner iwlist requested (Linux)')
    elif args.iw:
        scanner = IwWiFiScanner(interface=args.interface)
        LOGGER.info('- Scanner iw requested (Linux)')
    elif args.netsh:
        scanner = WindowsWiFiScanner(interface=args.interface)
        LOGGER.info('- Scanner netsh requested')
    else:
        if running_on_windows():
            LOGGER.info('- Windows Scanner netsh selected')
            scanner = WindowsWiFiScanner(args.interface)
        elif running_on_linux():
            if NetworkManagerWiFiScanner.is_available():
                scanner = NetworkManagerWiFiScanner(args.interface)
                # LOGGER.info('- Linux Scanner nmcli selected')
            elif IwlistWiFiScanner.is_available():
                scanner = IwlistWiFiScanner(args.interface)
                # LOGGER.info('- Linux Scanner iwlist selected')
            else:
                scanner = IwWiFiScanner(args.interface)
                # LOGGER.info('- Linux Scanner iw selected')
        else:
            LOGGER.critical('- OS not supported.')

    return scanner    

def identify_wifi_adapters() -> List[str]:
    """Return list of installed wifi adapers, None if not found"""
    adapters: List[str] = []
    if running_on_linux():
        cmd_output, ret_cd = ScannerBase._execute_process(CONSTANTS.IWCONFIG, show_feedback=False)
        if ret_cd == 0:
            for line in cmd_output:
                if 'ESSID' in line:
                    adapters.append(line.split()[0].strip())
    elif running_on_windows():
        cmd_output, ret_cd = ScannerBase._execute_process('netsh wlan show interfaces', show_feedback=False)
        if ret_cd == 0:
            for line in cmd_output:
                if line.strip().startswith('Name'):
                    adapters.append(line.split(':')[1].strip())

    if len(adapters) > 0:
        return adapters
    
    return None

def adapter_list() -> List[str]:
    """Return list of ALL installed network adapters"""
    # TODO: Build interface list
    adapters = []
    if running_on_linux():
        lines, _ = ScannerBase._execute_process(f'{CONSTANTS.IFCONFIG} -a', False)
        for line in lines:
             if 'flags' in line:
                 iface_name = line.split(':',1)[0].strip()
                 adapters.append(iface_name)
                 
    elif running_on_windows():
        lines, _ = ScannerBase._execute_process('ipconfig /all', False)
        for line in lines:
            if 'adapter' in line and '* ' not in line:
                iface_name = line.split('adapter')[1].replace(':','').strip()
                adapters.append(iface_name)
    else:
        pass # Unsupported OS

    LOGGER.debug(f'- adapters: {", ".join(adapters)}')
    return adapters


# ============================================================================================================================
# == Main Entrypoint =========================================================================================================   
def main() -> int:
    desc = 'Scan for wi-fi access points (Networks)'
    epilog = '''
This utility will scan for network APs (Access Points) using underlying OS utilities
and list related information.

- Supports Linux and Windows. 
- Output options include: formatted (default), csv and json

'''
    development_mode = False
    for arg in sys.argv:
        if arg == '-d':
            development_mode = True
            sys.argv.remove('-d')
            break

    parser = argparse.ArgumentParser(prog="ap_scanner", 
                                     description=desc, formatter_class=argparse.RawTextHelpFormatter,
                                     epilog=epilog)
    parser.add_argument('-i', '--interface', type=str, default=None, metavar='<iface>', help='Interface to use, default=first wireless adapter discovered')
    parser.add_argument('-v', '--verbose', action='count', default=0, help='Debug/verbose output to console')
    parser.add_argument('-j', '--json', action='store_true', default=False, help='Output json result')
    parser.add_argument('-c', '--csv', action='store_true', default=False, help='Output csv result')
    parser.add_argument('-r', '--rescan', action='store_true', default=False, help='(Windows only) force network rescan for APs')
    parser.add_argument('--nmcli', action='store_true', default=False, help='Force Linux Network Manager discovery')
    parser.add_argument('--iwlist', action='store_true', default=False, help='Force Linux iwlist discovery')
    parser.add_argument('--iw', action='store_true', default=False, help='Force Linux iw discover')
    parser.add_argument('--netsh', action='store_true', default=False, help='Force Windows netsh discovery')
    if development_mode:
        parser.add_argument('-t', '--test', type=str, default=None, metavar='<filename>', help='Use test data, specify filename')
        parser.add_argument('-s', '--save', type=str, default=None, metavar='<filename>', help='Filename to save (os scan) command output in')

    args = parser.parse_args()

    LOG_LVL = "INFO"
    if args.verbose == 1:
        LOG_LVL = "DEBUG"
    elif args.verbose > 1:
        LOG_LVL = "TRACE"

    # Remove root logger and create console logger
    LOGGER.remove(0) 
    h_console = LOGGER.add(sink=sys.stderr, level=LOG_LVL, format=CONSTANTS.CONSOLE_LOGFORMAT)
    ScannerBase.logging_level = LOG_LVL
    
    header_width = len(desc) + 20
    title = f'{parser.prog} v{version(parser.prog)}'.center(header_width-4, ' ')
    display_desc = desc.center(header_width-4, ' ')
    LOGGER.info('='*header_width)
    LOGGER.info(f'=={title}==')
    LOGGER.info('='*header_width)
    LOGGER.info(f'=={display_desc}==')
    LOGGER.info('='*header_width)
    LOGGER.info('')
    LOGGER.info('Validate command line options')
    
    if development_mode:
        LOGGER.warning('- Development mode enabled')
    else:
        # Disable development mode functionality
        args.test = False
        args.save = False

    wifi_adapters = identify_wifi_adapters()
    if wifi_adapters is None:
        LOGGER.critical('WiFi capabilities required. No Wifi adapter detected.  ABORT')
        return -1
    else:
        LOGGER.info(f'- {len(wifi_adapters)} Wifi adapter(s) detected: {", ".join(wifi_adapters)}')

    if args.interface:
        iface_list = adapter_list()
        if args.interface not in iface_list:
            LOGGER.error(f'- Invalid interface [{args.interface}], valid values: {", ".join(adapter_list())}')
            return -2
    else:
        args.interface = 'wlan0'
        if len(wifi_adapters) > 0:
            args.interface = wifi_adapters[0]

    wifi_adapter = AdapterInfo(args.interface)
    if wifi_adapter.connected:
        LOGGER.info(f'- "{wifi_adapter.name}" will be used to scan via {wifi_adapter.radio_type} [{wifi_adapter.signal}%]')
    else:
        LOGGER.info(f'- "{wifi_adapter.name}" will be used to scan')

    scanner = identify_scanner(args)
    if scanner is None:
        return -3
            
    if args.test:
        if not scanner.set_test_datafile(args.test):
            return -4
        elif args.rescan:
            LOGGER.warning('- TEST MODE: rescan otion ignored')
            args.rescan = False
        elif args.save:
            LOGGER.warning('- TEST MODE: save output option ignored')
            args.save = None
    else:
        # if not scanner.os_check():
        if not os_check(scanner):
            LOGGER.critical(f'Invalid scanner - {scanner.__class__.__name__} only valid for {scanner.scanner_supported_os()}')
            return -5

    if args.rescan:
        if not scanner.rescan():
            return -6
    
    if args.save: 
        scanner.set_output_capture_file(args.save)

    ap_list = scanner.scan_for_access_points()
    if ap_list is None or len(ap_list) == 0:
        LOGGER.error('No Access Points discovered. Process terminating...')
        return -99
    
    LOGGER.success(f'- {len(ap_list)} APs discovered')
    if args.json:
        display_json(ap_list)
    elif args.csv:
        display_csv(ap_list)
    else:
        display_access_points(ap_list)

    LOGGER.info('')
    return 0

if __name__ == "__main__":
    sys.exit(main())
 